drivers/sensor/lps22hh/lps22hh.c

```diff
@@ -207,38 +202,84 @@ static int lps22hh_init(const struct device *dev)
-	.drdy_flags	= DT_INST_GPIO_FLAGS(0, drdy_gpios),
-#endif
-#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
-	.bus_init = lps22hh_spi_init,
-	.spi_conf.frequency = DT_INST_PROP(0, spi_max_frequency),
-	.spi_conf.operation = (SPI_OP_MODE_MASTER | SPI_MODE_CPOL |
-			       SPI_MODE_CPHA | SPI_WORD_SET(8)),
-	.spi_conf.slave     = DT_INST_REG_ADDR(0),
-#if DT_INST_SPI_DEV_HAS_CS_GPIOS(0)
-	.gpio_cs_port	    = DT_INST_SPI_DEV_CS_GPIOS_LABEL(0),
-	.cs_gpio	    = DT_INST_SPI_DEV_CS_GPIOS_PIN(0),
-	.cs_gpio_flags	    = DT_INST_SPI_DEV_CS_GPIOS_FLAGS(0),
-
-	.spi_conf.cs        =  &lps22hh_data.cs_ctrl,
-#else
-	.spi_conf.cs        = NULL,
-#endif
-#elif DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
-	.bus_init = lps22hh_i2c_init,
-	.i2c_slv_addr = DT_INST_REG_ADDR(0),
+#define LPS22HH_CFG_IRQ(inst) \
+	.gpio_int = GPIO_DT_SPEC_INST_GET(inst, drdy_gpios),
 #else
-#error "BUS MACRO NOT DEFINED IN DTS"
-#endif
-};
+#define LPS22HH_CFG_IRQ(inst)
+#endif /* CONFIG_LPS22HH_TRIGGER */
+
+#define LPS22HH_SPI_OPERATION (SPI_WORD_SET(8) |			\
+				SPI_OP_MODE_MASTER |			\
+				SPI_MODE_CPOL |				\
+				SPI_MODE_CPHA)				\
+
+#define LPS22HH_CONFIG_SPI(inst)					\
+	{								\
+		.ctx = {						\
+			.read_reg =					\
+			   (stmdev_read_ptr) stmemsc_spi_read,		\
+			.write_reg =					\
+			   (stmdev_write_ptr) stmemsc_spi_write,	\
+			.handle =					\
+			   (void *)&lps22hh_config_##inst.stmemsc_cfg,	\
+		},							\
+		.stmemsc_cfg = {					\
+			.spi = SPI_DT_SPEC_INST_GET(inst,		\
+					   LPS22HH_SPI_OPERATION,	\
+					   0),				\
+		},							\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, drdy_gpios),	\
+			(LPS22HH_CFG_IRQ(inst)), ())			\
+	}
+
+/*
+ * Instantiation macros used when a device is on an I2C bus.
+ */
+
+#define LPS22HH_CONFIG_I2C(inst)					\
+	{								\
+		.ctx = {						\
+			.read_reg =					\
+			   (stmdev_read_ptr) stmemsc_i2c_read,		\
+			.write_reg =					\
+			   (stmdev_write_ptr) stmemsc_i2c_write,	\
+			.handle =					\
+			   (void *)&lps22hh_config_##inst.stmemsc_cfg,	\
+		},							\
+		.stmemsc_cfg = {					\
+			.i2c = I2C_DT_SPEC_INST_GET(inst),		\
+		},							\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, drdy_gpios),	\
+			(LPS22HH_CFG_IRQ(inst)), ())			\
+	}
+
+/*
+ * Main instantiation macro. Use of COND_CODE_1() selects the right
+ * bus-specific macro at preprocessor time.
+ */
 
-DEVICE_DT_INST_DEFINE(0, lps22hh_init, NULL,
-		    &lps22hh_data, &lps22hh_config, POST_KERNEL,
-		    CONFIG_SENSOR_INIT_PRIORITY, &lps22hh_api_funcs);
+#define LPS22HH_DEFINE(inst)						\
+	static struct lps22hh_data lps22hh_data_##inst;		\
+	static const struct lps22hh_config lps22hh_config_##inst =	\
+	COND_CODE_1(DT_INST_ON_BUS(inst, spi),				\
+		    (LPS22HH_CONFIG_SPI(inst)),			\
+		    (LPS22HH_CONFIG_I2C(inst)));			\
+	DEVICE_DT_INST_DEFINE(inst,					\
+			    lps22hh_init,				\
+			    NULL,					\
+			    &lps22hh_data_##inst,			\
+			    &lps22hh_config_##inst,			\
+			    POST_KERNEL,				\
+			    CONFIG_SENSOR_INIT_PRIORITY,		\
+			    &lps22hh_driver_api);
```

Comment:
non-blocking: I think this formatting is better:

```c
	DEVICE_DT_INST_DEFINE(inst, lps22hh_init, NULL, &lps22hh_data_##inst,  \
			      &lps22hh_config_##inst, POST_KERNEL,	       \
			      CONFIG_SENSOR_INIT_PRIORITY, &lps22hh_driver_api);
```

Reply:
You mean the less lines the better?

Reply:
yes

Reply:
ok, I followed your suggestion, even if I have no particular preference on one or the other.